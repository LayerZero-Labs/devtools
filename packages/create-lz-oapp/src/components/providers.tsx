import React, { useState } from "react"
import { QueryClient, QueryClientProvider } from "@tanstack/react-query"

export const Providers: React.FC<React.PropsWithChildren> = ({ children }) => {
    const [queryClient] = useState(() => new QueryClient())

    return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
}
