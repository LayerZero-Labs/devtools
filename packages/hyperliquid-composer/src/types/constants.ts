export const HYPERLIQUID_URLS = {
    MAINNET: 'https://api.hyperliquid.xyz',
    TESTNET: 'https://api.hyperliquid-testnet.xyz',
}

export const RPC_URLS = {
    MAINNET: 'https://rpc.hyperliquid.xyz/evm',
    TESTNET: 'https://rpc.hyperliquid-testnet.xyz/evm',
}

export const CHAIN_IDS = {
    MAINNET: 999,
    TESTNET: 998,
}

export const ENDPOINTS = {
    INFO: '/info',
    EXCHANGE: '/exchange',
}

export const MAX_HYPERCORE_SUPPLY = 2 ** 64 - 1

/**
 * These are the token ids for USDC on hypercore mainnet and testnet
 * It can be reproduced by grabbing the entire spotMeta and finding USDC because Hyperliquid does not have an API to target query an asset (April 8, 2025)
 *  -   curl -X POST "https://api.hyperliquid.xyz/info" -H "Content-Type: application/json" -d '{"type": "spotMeta"}' > spotOut.json
 *  -   Goto tokens and the first entry should be USDC with an index of 0
 */
export const USDC_TOKEN_ID = {
    MAINNET: 0,
    TESTNET: 0,
}

export function toAssetBridgeAddress(tokenIndex: number): string {
    const addressLength = 42
    const addressPrefix = '0x2'
    const indexAsHex = Number(tokenIndex).toString(16)

    const addressLengthWithoutPrefix = addressLength - addressPrefix.length
    return `${addressPrefix}${indexAsHex.padStart(addressLengthWithoutPrefix, '0')}`
}
