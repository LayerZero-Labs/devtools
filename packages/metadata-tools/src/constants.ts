export const METADATA_URL = process.env.LZ_METADATA_URL || 'https://metadata.layerzero-api.com/v1/metadata'
