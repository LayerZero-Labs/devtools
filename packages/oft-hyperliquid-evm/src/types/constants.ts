export const HYPERLIQUID_URLS = {
    MAINNET: 'https://api.hyperliquid.xyz',
    TESTNET: 'https://api.hyperliquid-testnet.xyz',
}

export const RPC_URLS = {
    MAINNET: 'https://rpc.hyperliquid.xyz/evm',
    TESTNET: 'https://rpc.hyperliquid-testnet.xyz/evm',
}

export const ENDPOINTS = {
    INFO: '/info',
    EXCHANGE: '/exchange',
}
