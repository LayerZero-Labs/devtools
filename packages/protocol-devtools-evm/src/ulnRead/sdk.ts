import type { IUlnRead, UlnReadUlnConfig, UlnReadUlnUserConfig } from '@layerzerolabs/protocol-devtools'
import { UlnReadUlnConfigSchema } from '@layerzerolabs/protocol-devtools'
import {
    OmniAddress,
    type OmniTransaction,
    compareBytes32Ascending,
    isDeepEqual,
    OmniPoint,
} from '@layerzerolabs/devtools'
import assert from 'assert'
import { printBoolean, printJson } from '@layerzerolabs/io-devtools'
import { isZero, AsyncRetriable } from '@layerzerolabs/devtools'
import { OmniSDK, Provider, addChecksum, makeZeroAddress } from '@layerzerolabs/devtools-evm'
import { Contract } from '@ethersproject/contracts'
// Although this SDK is not specific to SendUln302, it uses the SendUln302 ABI
// because it contains all the necessary method fragments
import { abi } from '@layerzerolabs/lz-evm-sdk-v2/artifacts/contracts/uln/readlib/ReadLib1002.sol/ReadLib1002.json'

export class UlnRead extends OmniSDK implements IUlnRead {
    constructor(provider: Provider, point: OmniPoint) {
        super({ eid: point.eid, contract: new Contract(point.address, abi).connect(provider) })
    }

    /**
     * @see {@link IUlnRead.getUlnConfig}
     */
    @AsyncRetriable()
    async getUlnConfig(channelId: number, address: OmniAddress | null | undefined): Promise<UlnReadUlnConfig> {
        this.logger.debug(`Getting ULN read config for eid ${channelId} and address ${makeZeroAddress(address)}`)

        const config = await this.contract.contract.getReadLibConfig(makeZeroAddress(address), channelId)
        // Now we convert the ethers-specific object into the common structure
        //
        // Here we need to spread the config into an object because what ethers gives us
        // is actually an array with extra properties
        return UlnReadUlnConfigSchema.parse({ ...config })
    }

    /**
     * @see {@link IUlnRead.getAppUlnConfig}
     */
    @AsyncRetriable()
    async getAppUlnConfig(channelId: number, address: OmniAddress): Promise<UlnReadUlnConfig> {
        this.logger.verbose(
            `Getting App ULN read config for channelId ${channelId} and address ${makeZeroAddress(address)}`
        )

        if (isZero(address)) {
            this.logger.warn(
                `Getting App ULN read config for channelId ${channelId} : Passed in OApp address is zero. This will request the default config.`
            )
        }

        const config = await this.contract.contract.getAppReadLibConfig(makeZeroAddress(address), channelId)

        // Now we convert the ethers-specific object into the common structure
        //
        // Here we need to spread the config into an object because what ethers gives us
        // is actually an array with extra properties
        return UlnReadUlnConfigSchema.parse({ ...config })
    }

    /**
     * @see {@link IUlnRead.hasAppUlnConfig}
     */
    async hasAppUlnConfig(channelId: number, oapp: string, config: UlnReadUlnUserConfig): Promise<boolean> {
        this.logger.verbose(`Checking whether ULN read configs for channelId ${channelId} and OApp ${oapp} match`)

        const currentConfig = await this.getAppUlnConfig(channelId, oapp)
        const currentSerializedConfig = this.serializeUlnConfig(currentConfig)
        const serializedConfig = this.serializeUlnConfig(config)

        this.logger.debug(`Current ULN read config: ${printJson(currentSerializedConfig)}`)
        this.logger.debug(`Incoming ULN read config: ${printJson(serializedConfig)}`)

        const areEqual = isDeepEqual(serializedConfig, currentSerializedConfig)

        return this.logger.verbose(`Checked ULN read configs: ${printBoolean(areEqual)}`), areEqual
    }

    decodeUlnConfig(ulnConfigBytes: string): UlnReadUlnConfig {
        const [rtnConfig] = this.contract.contract.interface.decodeFunctionResult('getReadLibConfig', ulnConfigBytes)

        return UlnReadUlnConfigSchema.parse({ ...rtnConfig })
    }

    encodeUlnConfig(config: UlnReadUlnUserConfig): string {
        const serializedConfig = this.serializeUlnConfig(config)
        const encoded = this.contract.contract.interface.encodeFunctionResult('getReadLibConfig', [serializedConfig])

        return assert(typeof encoded === 'string', 'Must be a string'), encoded
    }

    async setDefaultUlnConfig(channelId: number, config: UlnReadUlnUserConfig): Promise<OmniTransaction> {
        const serializedConfig = this.serializeUlnConfig(config)
        const data = this.contract.contract.interface.encodeFunctionData('setDefaultReadLibConfigs', [
            [
                {
                    eid: channelId,
                    config: serializedConfig,
                },
            ],
        ])

        return {
            ...this.createTransaction(data),
            description: `Setting default ULN config for ${channelId}: ${printJson(serializedConfig)}`,
        }
    }

    /**
     * Prepares the ULN config to be sent to the contract
     *
     * This involves adding two properties that are required by the EVM
     * contracts (for optimization purposes) but don't need to be present
     * in our configuration and ensuring correct checksum on the DVN addresses.
     *
     * @param {UlnReadUlnUserConfig} config
     * @returns {SerializedUlnReadUlnConfig}
     */
    protected serializeUlnConfig({
        requiredDVNs,
        optionalDVNs = [],
        optionalDVNThreshold = 0,
        executor = makeZeroAddress(),
    }: UlnReadUlnUserConfig): SerializedUlnReadUlnConfig {
        return {
            executor,
            requiredDVNCount: requiredDVNs.length,
            optionalDVNCount: optionalDVNs.length,
            optionalDVNThreshold,
            requiredDVNs: requiredDVNs.map(addChecksum).sort(compareBytes32Ascending),
            optionalDVNs: optionalDVNs.map(addChecksum).sort(compareBytes32Ascending),
        }
    }
}

/**
 * Helper type that matches the expected UlnConfig type for the solicity implementation
 */
interface SerializedUlnReadUlnConfig extends UlnReadUlnConfig {
    requiredDVNCount: number
    optionalDVNCount: number
}
