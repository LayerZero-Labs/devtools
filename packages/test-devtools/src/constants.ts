import { ChannelId, EndpointId } from '@layerzerolabs/lz-definitions'
import { wordlist } from '@scure/bip39/wordlists/english'

export const ENDPOINT_IDS = Object.values(EndpointId).filter((value): value is EndpointId => typeof value === 'number')

export const CHANNEL_IDS = Object.values(ChannelId).filter((value): value is ChannelId => typeof value === 'number')

export const BIP39_WORDLIST: string[] = wordlist
