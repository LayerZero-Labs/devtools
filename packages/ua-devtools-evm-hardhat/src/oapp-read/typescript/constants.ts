export const READ_LIBRARY: string = 'readLibrary'
export const READ_CHANNEL_CONFIGS: string = 'readChannelConfigs'
export const CHANNEL_ID: string = 'channelId'
export const ACTIVE: string = 'active'
