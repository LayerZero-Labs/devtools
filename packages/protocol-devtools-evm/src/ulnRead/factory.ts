import pMemoize from 'p-memoize'
import type { UlnReadFactory } from '@layerzerolabs/protocol-devtools'
import { UlnRead } from './sdk'
import type { OmniPoint } from '@layerzerolabs/devtools'
import type { ProviderFactory } from '@layerzerolabs/devtools-evm'

/**
 * Syntactic sugar that creates an instance of EVM `UlnRead` SDK
 * based on an `OmniPoint` with help of an `OmniContractFactory`
 *
 * @param {ProviderFactory} providerFactory
 * @returns {UlnReadFactory<UlnRead>}
 */
export const createUlnReadFactory = (providerFactory: ProviderFactory): UlnReadFactory<UlnRead, OmniPoint> =>
    pMemoize(async (point) => new UlnRead(await providerFactory(point.eid), point))
