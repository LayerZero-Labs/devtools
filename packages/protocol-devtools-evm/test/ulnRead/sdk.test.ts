import { addChecksum, makeZeroAddress, Provider } from '@layerzerolabs/devtools-evm'
import { MainnetV2EndpointId } from '@layerzerolabs/lz-definitions'
import { AddressZero } from '@ethersproject/constants'
import { UlnReadUlnConfig } from '@layerzerolabs/protocol-devtools'
import fc from 'fast-check'
import { channelIdArbitrary, evmAddressArbitrary } from '@layerzerolabs/test-devtools'
import { compareBytes32Ascending } from '@layerzerolabs/devtools'
import { UlnReadUlnUserConfig } from '@layerzerolabs/protocol-devtools'
import { JsonRpcProvider } from '@ethersproject/providers'
import { UlnRead } from '@/ulnRead'

jest.spyOn(JsonRpcProvider.prototype, 'detectNetwork').mockResolvedValue({ chainId: 1, name: 'mock' })

describe('ulnRead/sdk', () => {
    let provider: Provider, ulnSdk: UlnRead

    const dvnsArbitrary = fc.array(evmAddressArbitrary, { minLength: 2 })

    beforeEach(async () => {
        provider = new JsonRpcProvider()
        ulnSdk = new UlnRead(provider, { eid: MainnetV2EndpointId.ETHEREUM_V2_MAINNET, address: makeZeroAddress() })
    })

    describe('encodeUlnConfig', () => {
        it('should encode and decode the UlnReadUlnConfig', async () => {
            const ulnConfig: UlnReadUlnConfig = {
                executor: AddressZero,
                optionalDVNThreshold: 1,
                optionalDVNs: [AddressZero, AddressZero],
                requiredDVNs: [AddressZero],
            }
            const ulnConfigEncoded = ulnSdk.encodeUlnConfig(ulnConfig)
            expect(ulnConfigEncoded).toMatchSnapshot()

            const ulnConfigDecoded: UlnReadUlnConfig = ulnSdk.decodeUlnConfig(ulnConfigEncoded)

            expect(ulnConfig).toEqual(ulnConfigDecoded)
        })

        it('should sort requiredDVNs and optionalDVNs before encoding', async () => {
            fc.assert(
                fc.property(dvnsArbitrary, (dvns) => {
                    const sortedDvns = [...dvns].sort((a, b) => a.localeCompare(b))
                    const ulnConfigUnsorted: UlnReadUlnConfig = {
                        executor: AddressZero,
                        optionalDVNThreshold: 0,
                        optionalDVNs: dvns,
                        requiredDVNs: dvns,
                    }

                    const ulnConfigSorted: UlnReadUlnConfig = {
                        ...ulnConfigUnsorted,
                        optionalDVNs: sortedDvns,
                        requiredDVNs: sortedDvns,
                    }

                    const ulnConfigEncodedUnsorted = ulnSdk.encodeUlnConfig(ulnConfigUnsorted)
                    const ulnConfigEncodedSorted = ulnSdk.encodeUlnConfig(ulnConfigSorted)
                    expect(ulnConfigEncodedSorted).toBe(ulnConfigEncodedUnsorted)
                }),
                { numRuns: 20 }
            )
        })
    })

    describe('setDefaultUlnConfig', () => {
        it('should sort requiredDVNs and optionalDVNs', async () => {
            await fc.assert(
                fc.asyncProperty(channelIdArbitrary, dvnsArbitrary, async (eid, dvns) => {
                    const sortedDvns = [...dvns].sort((a, b) => a.localeCompare(b))
                    const ulnConfigUnsorted: UlnReadUlnConfig = {
                        executor: AddressZero,
                        optionalDVNThreshold: 0,
                        optionalDVNs: dvns,
                        requiredDVNs: dvns,
                    }

                    const ulnConfigSorted: UlnReadUlnConfig = {
                        ...ulnConfigUnsorted,
                        optionalDVNs: sortedDvns,
                        requiredDVNs: sortedDvns,
                    }

                    // Let's check that both the sorted and the unsorted config produce the same transaction
                    const transactionUnsorted = await ulnSdk.setDefaultUlnConfig(eid, ulnConfigUnsorted)
                    const transactionsSorted = await ulnSdk.setDefaultUlnConfig(eid, ulnConfigSorted)
                    expect(transactionUnsorted).toEqual(transactionsSorted)

                    // And let's check that the encoding call is correct and the DVNs are sorted
                    expect(transactionsSorted.data).toBe(
                        ulnSdk.contract.contract.interface.encodeFunctionData('setDefaultReadLibConfigs', [
                            [
                                {
                                    eid,
                                    config: {
                                        executor: AddressZero,
                                        optionalDVNThreshold: 0,
                                        optionalDVNs: sortedDvns.map(addChecksum),
                                        requiredDVNs: sortedDvns.map(addChecksum),
                                        requiredDVNCount: sortedDvns.length,
                                        optionalDVNCount: sortedDvns.length,
                                    },
                                },
                            ],
                        ])
                    )
                }),
                { numRuns: 20 }
            )
        })

        it('should default the missing attributes', async () => {
            await fc.assert(
                fc.asyncProperty(channelIdArbitrary, dvnsArbitrary, async (eid, dvns) => {
                    const sortedDvns = [...dvns].sort((a, b) => a.localeCompare(b))
                    const ulnUserConfig: UlnReadUlnUserConfig = {
                        requiredDVNs: dvns,
                    }
                    const ulnConfig: UlnReadUlnConfig = {
                        requiredDVNs: dvns,
                        executor: AddressZero,
                        optionalDVNThreshold: 0,
                        optionalDVNs: [],
                    }

                    // Let's check that both the sorted and the unsorted config produce the same transaction
                    const transactionsWithDefaults = await ulnSdk.setDefaultUlnConfig(eid, ulnUserConfig)
                    const transactionsWithFullConfig = await ulnSdk.setDefaultUlnConfig(eid, ulnConfig)
                    expect(transactionsWithDefaults).toEqual(transactionsWithFullConfig)

                    // And let's check that the encoding call is correct and the DVNs are sorted
                    expect(transactionsWithDefaults.data).toBe(
                        ulnSdk.contract.contract.interface.encodeFunctionData('setDefaultReadLibConfigs', [
                            [
                                {
                                    eid,
                                    config: {
                                        executor: ulnConfig.executor,
                                        optionalDVNThreshold: ulnConfig.optionalDVNThreshold,
                                        optionalDVNs: ulnConfig.optionalDVNs,
                                        requiredDVNs: sortedDvns.map(addChecksum),
                                        requiredDVNCount: ulnConfig.requiredDVNs.length,
                                        optionalDVNCount: ulnConfig.optionalDVNs.length,
                                    },
                                },
                            ],
                        ])
                    )
                }),
                { numRuns: 20 }
            )
        })
    })

    describe('hasAppUlnConfig()', () => {
        const ulnConfigArbitrary = dvnsArbitrary.chain((dvns) =>
            fc.record<UlnReadUlnConfig>({
                executor: evmAddressArbitrary,
                optionalDVNThreshold: fc.integer({ min: 0, max: dvns.length }),
                optionalDVNs: fc.constant(dvns),
                requiredDVNs: fc.constant(dvns),
            })
        )

        let getAppUlnConfigSpy: jest.SpyInstance<Promise<UlnReadUlnConfig>, [channelId: number, address: string]>

        beforeEach(() => {
            getAppUlnConfigSpy = jest.spyOn(UlnRead.prototype, 'getAppUlnConfig')
        })

        afterEach(() => {
            getAppUlnConfigSpy.mockRestore()
        })

        it('should return true if the configs are identical', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    async (channelId, oapp, ulnConfig) => {
                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(ulnSdk.hasAppUlnConfig(channelId, oapp, ulnConfig)).resolves.toBeTruthy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return true if config has extra properties', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    fc.object(),
                    async (channelId, oapp, ulnConfig, extra) => {
                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(
                            ulnSdk.hasAppUlnConfig(channelId, oapp, {
                                ...extra,
                                ...ulnConfig,
                            })
                        ).resolves.toBeTruthy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return true if configs with defaults are identical', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    dvnsArbitrary,
                    async (channelId, oapp, dvns) => {
                        const ulnUserConfig: UlnReadUlnUserConfig = {
                            requiredDVNs: dvns,
                        }
                        const ulnConfig: UlnReadUlnConfig = {
                            requiredDVNs: dvns,
                            optionalDVNs: [],
                            optionalDVNThreshold: 0,
                            executor: AddressZero,
                        }

                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(ulnSdk.hasAppUlnConfig(channelId, oapp, ulnUserConfig)).resolves.toBeTruthy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return false if configs with defaults are not identical', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    async (channelId, oapp, ulnConfig) => {
                        // We only want to test againsts configs that are not
                        // equal to the default config
                        fc.pre(
                            ulnConfig.executor !== AddressZero ||
                                ulnConfig.optionalDVNThreshold !== 0 ||
                                ulnConfig.optionalDVNs.length !== 0
                        )

                        const ulnUserConfig: UlnReadUlnUserConfig = {
                            requiredDVNs: ulnConfig.requiredDVNs,
                        }

                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(ulnSdk.hasAppUlnConfig(channelId, oapp, ulnUserConfig)).resolves.toBeFalsy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return true if the configs are identical except for the dvn casing', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    async (channelId, oapp, ulnConfig) => {
                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(
                            ulnSdk.hasAppUlnConfig(channelId, oapp, {
                                ...ulnConfig,
                                requiredDVNs: ulnConfig.requiredDVNs.map(addChecksum),
                                optionalDVNs: ulnConfig.optionalDVNs.map(addChecksum),
                            })
                        ).resolves.toBeTruthy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return true if the configs are identical except for the dvn sorting and casing', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    async (channelId, oapp, ulnConfig) => {
                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(
                            ulnSdk.hasAppUlnConfig(channelId, oapp, {
                                ...ulnConfig,
                                requiredDVNs: ulnConfig.requiredDVNs.map(addChecksum).sort(compareBytes32Ascending),
                                optionalDVNs: ulnConfig.optionalDVNs.map(addChecksum).sort(compareBytes32Ascending),
                            })
                        ).resolves.toBeTruthy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return false if the confirmations are different', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    evmAddressArbitrary,
                    async (channelId, oapp, ulnConfig, executor) => {
                        fc.pre(executor !== ulnConfig.executor)

                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(
                            ulnSdk.hasAppUlnConfig(channelId, oapp, {
                                ...ulnConfig,
                                executor,
                            })
                        ).resolves.toBeFalsy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return false if the optionalDVNThresholds are different', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    fc.integer(),
                    async (channelId, oapp, ulnConfig, optionalDVNThreshold) => {
                        fc.pre(optionalDVNThreshold !== ulnConfig.optionalDVNThreshold)

                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(
                            ulnSdk.hasAppUlnConfig(channelId, oapp, {
                                ...ulnConfig,
                                optionalDVNThreshold,
                            })
                        ).resolves.toBeFalsy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return false if the requiredDVNs are different', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    dvnsArbitrary,
                    async (channelId, oapp, ulnConfig, extraRequiredDVNs) => {
                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(
                            ulnSdk.hasAppUlnConfig(channelId, oapp, {
                                ...ulnConfig,
                                requiredDVNs: [...ulnConfig.requiredDVNs, ...extraRequiredDVNs],
                            })
                        ).resolves.toBeFalsy()
                    }
                ),
                { numRuns: 20 }
            )
        })

        it('should return false if the optionalDVNs are different', async () => {
            await fc.assert(
                fc.asyncProperty(
                    channelIdArbitrary,
                    evmAddressArbitrary,
                    ulnConfigArbitrary,
                    dvnsArbitrary,
                    async (channelId, oapp, ulnConfig, extraOptionalDVNs) => {
                        getAppUlnConfigSpy.mockReset()
                        getAppUlnConfigSpy.mockResolvedValue(ulnConfig)

                        await expect(
                            ulnSdk.hasAppUlnConfig(channelId, oapp, {
                                ...ulnConfig,
                                optionalDVNs: [...ulnConfig.optionalDVNs, ...extraOptionalDVNs],
                            })
                        ).resolves.toBeFalsy()
                    }
                ),
                { numRuns: 20 }
            )
        })
    })
})
