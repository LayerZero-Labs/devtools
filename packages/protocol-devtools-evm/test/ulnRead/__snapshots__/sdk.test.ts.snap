// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ulnRead/sdk encodeUlnConfig should encode and decode the UlnReadUlnConfig 1`] = `"0x0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000c0000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"`;
